pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,187 @@
+package server
+
+import (
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+type tagService struct {
+	distribution.TagService
+
+	repo *repository
+}
+
+func (t tagService) Get(ctx context.Context, tag string) (distribution.Descriptor, error) {
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return distribution.Descriptor{}, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	te := imageapi.LatestTaggedImage(imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	if !t.repo.pullthrough {
+		image, err := t.repo.getImage(dgst)
+		if err != nil {
+			return distribution.Descriptor{}, err
+		}
+
+		if !isImageManaged(image) {
+			return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+		}
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if !t.repo.pullthrough {
+			dgst, err := digest.ParseDigest(history.Items[0].Image)
+			if err != nil {
+				return nil, err
+			}
+
+			image, err := t.repo.getImage(dgst)
+			if err != nil {
+				return nil, err
```

Comment:
Skip `NotFoundErr`. Also in this case I'd just log the error (if not `NotFound`) and continue.
The same for other `getImage` occurences.